From 469fcb0ff9336acd5a9e6a5e305eadf60c9a67e0 Mon Sep 17 00:00:00 2001
From: kitakar5525 <34676735+kitakar5525@users.noreply.github.com>
Date: Thu, 21 Nov 2019 10:53:21 +0900
Subject: [PATCH 1/4] mwifiex: s0ix test

This patch is from sebanc; trying to fix the following issues:
- wifi crash after suspend when mwifiex modules remain loaded during suspend
- s0ix achievement when mwifiex modules remain loaded during suspend

The function mwifiex_pcie_suspend2 is from mwifiex_pcie_remove and
the function mwifiex_pcie_resume2 is largely from mwifiex_pcie_probe.
So, it behaves in the same way as we unload modules before suspend
and load modules after suspend.
Also disables bridge_d3 regarding wifi.

Link to the comment where sebanc posted this patch:
https://github.com/jakeday/linux-surface/issues/420#issuecomment-544615371

---
 drivers/net/wireless/marvell/mwifiex/pcie.c   | 78 ++++++++++++++++++-
 .../net/wireless/marvell/mwifiex/sta_cmd.c    |  2 +-
 2 files changed, 77 insertions(+), 3 deletions(-)

diff --git a/drivers/net/wireless/marvell/mwifiex/pcie.c b/drivers/net/wireless/marvell/mwifiex/pcie.c
index 76e76a6a7..7f4e75fa7 100644
--- a/drivers/net/wireless/marvell/mwifiex/pcie.c
+++ b/drivers/net/wireless/marvell/mwifiex/pcie.c
@@ -217,6 +217,76 @@ static int mwifiex_pcie_resume(struct device *dev)
 
 	return 0;
 }
+
+static int mwifiex_pcie_resume2(struct device *dev)
+{
+	struct pci_dev *pdev = to_pci_dev(dev);
+	struct pcie_service_card *card;
+	int ret;
+
+	pr_debug("info: vendor=0x%4.04X device=0x%4.04X rev=%d\n",
+		 pdev->vendor, pdev->device, pdev->revision);
+
+	card = pci_get_drvdata(pdev);
+
+	init_completion(&card->fw_done);
+
+	card->dev = pdev;
+
+	/* device tree node parsing and platform specific configuration*/
+	if (pdev->dev.of_node) {
+		ret = mwifiex_pcie_probe_of(&pdev->dev);
+		if (ret)
+			return ret;
+	}
+
+	if (mwifiex_add_card(card, &card->fw_done, &pcie_ops,
+			     MWIFIEX_PCIE, &pdev->dev)) {
+		pr_err("%s failed\n", __func__);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int mwifiex_pcie_suspend2(struct device *dev)
+{
+	struct pcie_service_card *card;
+	struct pci_dev *pdev = to_pci_dev(dev);
+	struct mwifiex_adapter *adapter;
+	struct mwifiex_private *priv;
+	const struct mwifiex_pcie_card_reg *reg;
+	u32 fw_status;
+	int ret;
+
+	card = pci_get_drvdata(pdev);
+
+	wait_for_completion(&card->fw_done);
+
+	adapter = card->adapter;
+	if (!adapter || !adapter->priv_num)
+		return 0;
+
+	reg = card->pcie.reg;
+	if (reg)
+		ret = mwifiex_read_reg(adapter, reg->fw_status, &fw_status);
+	else
+		fw_status = -1;
+
+	if (fw_status == FIRMWARE_READY_PCIE && !adapter->mfg_mode) {
+		mwifiex_deauthenticate_all(adapter);
+
+		priv = mwifiex_get_priv(adapter, MWIFIEX_BSS_ROLE_ANY);
+
+		mwifiex_disable_auto_ds(priv);
+
+		mwifiex_init_shutdown_fw(priv, MWIFIEX_FUNC_SHUTDOWN);
+	}
+
+	mwifiex_remove_card(adapter);
+
+	return 0;
+}
 #endif
 
 /*
@@ -267,6 +337,8 @@ static int mwifiex_pcie_probe(struct pci_dev *pdev,
 		return -1;
 	}
 
+	pdev->bus->self->bridge_d3 = false;
+
 	return 0;
 }
 
@@ -414,8 +486,10 @@ static const struct pci_error_handlers mwifiex_pcie_err_handler = {
 
 #ifdef CONFIG_PM_SLEEP
 /* Power Management Hooks */
-static SIMPLE_DEV_PM_OPS(mwifiex_pcie_pm_ops, mwifiex_pcie_suspend,
-				mwifiex_pcie_resume);
+//static SIMPLE_DEV_PM_OPS(mwifiex_pcie_pm_ops, mwifiex_pcie_suspend,
+//				mwifiex_pcie_resume);
+static SIMPLE_DEV_PM_OPS(mwifiex_pcie_pm_ops, mwifiex_pcie_suspend2,
+				mwifiex_pcie_resume2);
 #endif
 
 /* PCI Device Driver */
diff --git a/drivers/net/wireless/marvell/mwifiex/sta_cmd.c b/drivers/net/wireless/marvell/mwifiex/sta_cmd.c
index 485360e85..ab16c2d75 100644
--- a/drivers/net/wireless/marvell/mwifiex/sta_cmd.c
+++ b/drivers/net/wireless/marvell/mwifiex/sta_cmd.c
@@ -2395,7 +2395,7 @@ int mwifiex_sta_init_cmd(struct mwifiex_private *priv, u8 first_sta, bool init)
 	if (ret)
 		return -1;
 
-	if (!disable_auto_ds && first_sta &&
+	if (0 && !disable_auto_ds && first_sta &&
 	    priv->bss_type != MWIFIEX_BSS_TYPE_UAP) {
 		/* Enable auto deep sleep */
 		auto_ds.auto_ds = DEEP_SLEEP_ON;
-- 
2.24.0

